docs: Add example playbook and "readme" for confined users

vmojzis opened the pull request:
Note: Relabeling user home dirs should probably be added to https://docs.ansible.com/ansible/latest/collections/community/general/selogin_module.html

richm commented:
Just a few minor ansible-lint issues, otherwise, lgtm
would like @bachradsusi to review for correctness

spetrosi commented:
Why not include this into the main README.md?

bachradsusi commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,107 @@
+        comment: Linux System Roles SELinux User
+        name: "{{ item }}"
+        seuser: staff_u
+        group: "wheel"
```
I'd say that `wheel` group is not necessary given that all privileged users have their own sudo config line.

bachradsusi replied:
I'd drop creating users completely, expect that users are already created/available and use the role to create new login mappings - https://github.com/linux-system-roles/selinux/blob/main/README.md#selinux_logins

vmojzis replied:
It's an example, which is why I'm trying to show both ways. Anyone using the playbook can just set `privileged_users: []`.

bachradsusi commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,107 @@
+    # List of privileged users to be created
+    privileged_users:
+      - "staff"
+      - "roger"
```
README uses "mark" and "roger"

bachradsusi commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,107 @@
+      #- {login: '__default__', seuser: 'staff_u', serange: 's0-s0:c0.c1023', state: 'present'}
+      # Map admin users to SElinux user "sysadm_u"
+      - {login: 'root', seuser: 'sysadm_u', serange: 's0-s0:c0.c1023',
+               state: 'present'}
```
In "mls" we use "root" SELinux user for "root" user and `sysadm_u` for admin non-root users. I'd follow it here as well.

vmojzis replied:
Ok, I guess `root` can stay "unconfined" since it's not accessible from outside. I'll replace it with a better example.

bachradsusi commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,107 @@
+    - name: Relabel homedirs of users with new login mapping
+      shell: "restorecon -RF {{ item.stdout }}"
+      when: item.stdout != ""
+      with_items: "{{ __login_homedirs.results if __login_homedirs.results is iterable else [] }}"
```
Would it be possible to use https://github.com/linux-system-roles/selinux/blob/main/README.md#selinux_restore_dirs instead of `restorecon` command?

vmojzis replied:
This command needs to take place after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml.

vmojzis commented:
> Why not include this into the main README.md?

The intention was to not "force" people who only want a simple way for confining users to read the whole "readme" (`confined_users-playbook.yml` uses a subset of SELinux system role's capabilities). Also, `privileged_users` variable is specific to  `confined_users-playbook.yml`.

spetrosi commented:
> > Why not include this into the main README.md?
> 
> The intention was to not "force" people who only want a simple way for confining users to read the whole "readme" (`confined_users-playbook.yml` uses a subset of SELinux system role's capabilities). Also, `privileged_users` variable is specific to `confined_users-playbook.yml`.

The wording of `You can greatly improve security of your systems` sounds to me like a best practice for users to do. We can also restructure the README.md to split it into "scenarios" one of which would be `Configuring confined users`, like I did in mssql role https://linux-system-roles.github.io/mssql/. @bachradsusi WDYT?
We can do this in a future PR though, after the section is merged here as a separate readme.

bachradsusi commented:
Looking at https://linux-system-roles.github.io/selinux/ and https://linux-system-roles.github.io/mssql/ I guess it makes sense to merge "README-confined_users.md" into "README.md" as a new scenario.

Generally I like the idea to have confined users documented, but I think that the current approach unnecessarily mixes everything into one playbook. Also using `privileged_users` in the README is confusing for me as it's not related to the role, it's only implemented in the example playbook.

In the example playbook I'd also avoid all tasks which could be done by the role:

- instead of `shell: restorecon ...` use `selinux_restore_dirs` variable
- do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`

btw `user` module would not change SELinux user mappings if the user already exists

```
# ansible -i localhost, -c local -m user -a 'name=user_u seuser=user_u' all
localhost | CHANGED => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": true,
    "comment": "",
    "create_home": true,
    "group": 1002,
    "home": "/home/user_u",
    "name": "user_u",
    "shell": "/bin/bash",
    "state": "present",
    "system": false,
    "uid": 1002
}
# semanage login -l | grep user_u
user_u               user_u               s0                   *
# ansible -i localhost, -c local -m user -a 'name=user_u seuser=unconfined_u' all
localhost | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "append": false,
    "changed": false,
    "comment": "",
    "group": 1002,
    "home": "/home/user_u",
    "move_home": false,
    "name": "user_u",
    "shell": "/bin/bash",
    "state": "present",
    "uid": 1002
}
# semanage login -l | grep user_u
user_u               user_u               s0                   *
```

vmojzis commented:
> Looking at https://linux-system-roles.github.io/selinux/ and https://linux-system-roles.github.io/mssql/ I guess it makes sense to merge "README-confined_users.md" into "README.md" as a new scenario.
> 
> Generally I like the idea to have confined users documented, but I think that the current approach unnecessarily mixes everything into one playbook. Also using `privileged_users` in the README is confusing for me as it's not related to the role, it's only implemented in the example playbook.

It felt better to me to document the variable in a separate README than to leave it out. Yes, it is only used in the example playbook, but is important for anyone utilizing the example.

> 
> In the example playbook I'd also avoid all tasks which could be done by the role:

> 
> * instead of `shell: restorecon ...` use `selinux_restore_dirs` variable

`restorecon` needs to be run after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml for this to work properly.

> * do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`
> 

Why? This is an example so I feel it is important to show multiple approaches.

> btw `user` module would not change SELinux user mappings if the user already existed
> 
> ```
> # ansible -i localhost, -c local -m user -a 'name=user_u seuser=user_u' all
> localhost | CHANGED => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python3"
>     },
>     "changed": true,
>     "comment": "",
>     "create_home": true,
>     "group": 1002,
>     "home": "/home/user_u",
>     "name": "user_u",
>     "shell": "/bin/bash",
>     "state": "present",
>     "system": false,
>     "uid": 1002
> }
> # semanage login -l | grep user_u
> user_u               user_u               s0                   *
> # ansible -i localhost, -c local -m user -a 'name=user_u seuser=unconfined_u' all
> localhost | SUCCESS => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python3"
>     },
>     "append": false,
>     "changed": false,
>     "comment": "",
>     "group": 1002,
>     "home": "/home/user_u",
>     "move_home": false,
>     "name": "user_u",
>     "shell": "/bin/bash",
>     "state": "present",
>     "uid": 1002
> }
> # semanage login -l | grep user_u
> user_u               user_u               s0                   *
> ```

bachradsusi commented:
> It felt better to me to document the variable in a separate README than to leave it out. Yes, it is only used in the example playbook, but is important for anyone utilizing the example.

It depends on what's the goal. If the goal a generic SELinux example using Ansible SELinux modules and creating users then it's probably ok. But I'd expect that README and examples in linux-system-roles would be related to roles. I understand you need to create sudoers so that administrator users could switch to sysadm_t using sudo, but I don't think it's necessary to complicate example playbook with a new list variable and use loops. It would be enough to demonstrate the change on one hardcoded user which is documented as pre-created.
 
> > In the example playbook I'd also avoid all tasks which could be done by the role:
> 
> > * instead of `shell: restorecon ...` use `selinux_restore_dirs` variable
> 
> `restorecon` needs to be run after the login mapping is applied. So we'd need to change order of tasks in tasks/main.yml for this to work properly.

You are right. It's a bug in the role and the order needs to be changed. restorecon should be the last step.


> > * do not create users with assigned SELinux users, if you need to create a user, do it without SELinux part and demonstrate change of SELinux login mappings using `selinux_logins`
> 
> Why? This is an example so I feel it is important to show multiple approaches.

It also depends on the goal. Using `user` module with `seuser` option is not related to the role.


Anyway, even though I don't agree with some parts of this PR, I know that @vmojzis is experienced SELinux engineer and so if other maintainers approve this PR I'm not going to block it.

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+      # Map all new users to SElinux user "user_u" (instead of "unconfined_u")
+      - {login: '__default__', seuser: 'user_u', serange: 's0',
+         state: 'present'}
+      # Alternatively map all new users to SElinux user "staff_u" (instead of
```
```suggestion
        # Alternatively map all new users to SElinux user "staff_u" (instead of
```

vmojzis replied:
I skipped this one since it would indent the comment from the surrounding lines.

richm replied:
ok - what do you suggest we do to make yamllint happy?

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+      - {login: '__default__', seuser: 'user_u', serange: 's0',
+         state: 'present'}
+      # Alternatively map all new users to SElinux user "staff_u" (instead of
+      # "unconfined_u")
```
```suggestion
        # "unconfined_u")
```

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+      # Map select non-admin privileged users to SELinux user "staff_u"
+      # - {login: '<existing_user>', seuser: 'staff_u', serange: 's0',
+      #    state: 'present'}
+      #- {login: 'Lizzy', seuser: 'staff_u', serange: 's0', state: 'present'}
```
```suggestion
      # - {login: 'Lizzy', seuser: 'staff_u', serange: 's0', state: 'present'}
```

vmojzis replied:
Fixed, thanks.

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+  # Prepare the prerequisites required for this playbook
+  tasks:
+    # Enable kiosk mode -- add extremely limited users "guest" and "xguest"
+    #- name: Install xguest package
```
```suggestion
    # - name: Install xguest package
```

vmojzis replied:
Fixed, thanks.

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+    #    state: present
+
+    # Create login mappings for "existing_privileged_users"
+    - set_fact:
```
```suggestion
    - name: Build list of selinux logins
      set_fact:
```

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+
+    # Add new system users that will be assigned to staff_u SELinux user
+    - name: Add non-root privileged users
+      user:
```
I guess this assumes you aren't using IdM or some other system that you must create users in

vmojzis replied:
Yes, that seemed like too complex a task to have in an example.

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+        - name: Wait for managed host to come back
+          wait_for_connection:
+            delay: 10
+            timeout: 300
```
why so long?  Is this because selinux has to do some sort of filesystem-wide restorecon?

vmojzis replied:
This is just copied from the first example.

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+      shell: >
+        getent passwd "{{ item.login }}" | awk -F: '{ print $6 }'
+      register: __login_homedirs
+      with_items: "{{ selinux_logins }}"
```
```suggestion
      with_items: "{{ selinux_logins }}"
      changed_when: false
```

richm commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+    # relabeled.
+    # Gather homedirs of each user with a new login mapping
+    - name: Gather homedirs of users with new login mapping
+      shell: >
```
```suggestion
      shell: >-
        set -euo pipefail;
```

vmojzis replied:
Fixed, thanks (aside from the "-o pipefail").

richm commented:
> You are right. It's a bug in the role and the order needs to be changed. restorecon should be the last step.

Does this mean the role can't manage confined users until this is fixed?

bachradsusi commented:
> Does this mean the role can't manage confined users until this is fixed?

The user part of the file context - unconfined_u, staff_u, ... - is not used for access decisions in Fedora policies so `restorecon -F` is not actually necessary. 

```
$ sestatus  | grep policy\ name
Loaded policy name:             targeted
$ id -Z
user_u:user_r:user_t:s0
$ ls -Z date
unconfined_u:object_r:user_home_t:s0 date
$ cat date
Tue Aug 29 09:59:34 AM EDT 2023
```

It could trigger security scan warnings though.

vmojzis commented:
Sorry for the delay.
I dropped the use of `user` module, fixed relabeling (it is now the last task in the role and the example uses `selinux_restore_dirs`) and incorporated all the stylistic tweaks.

vmojzis commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+         state: 'present'}
+      # Alternatively map all new users to SElinux user "staff_u" (instead of
+      # "unconfined_u")
+      #- {login: '__default__', seuser: 'staff_u', serange: 's0-s0:c0.c1023',
```
Fixed, thanks.

vmojzis commented on examples/confined_users-playbook.yml:
```diff
@@ -0,0 +1,129 @@
+    # Relabel the gathered homedirs (-F is necessary to change the user portion
+    # of file labels)
+    - name: Relabel homedirs of users with new login mapping
+      shell: "restorecon -RF {{ item.stdout }}"
```
Removed (I used `selinux_restore_dirs` instead ).

richm commented on README-confined_users.md:
```diff
@@ -0,0 +1,59 @@
+# Confined users
```
```suggestion
# Confined users

```

vmojzis replied:
Fixed, thank you.

richm commented on README-confined_users.md:
```diff
@@ -0,0 +1,59 @@
+
+## Variables
+
+### selinux_booleans
```
```suggestion
### selinux_booleans

```

vmojzis replied:
Fixed, thank you.

richm commented:
@bachradsusi is this PR ok to merge?

vmojzis commented:
Fixed some ansible-lint issues (but I cannot fix the "pipefail" and do not agree with the comment indentation on line 44).

bachradsusi commented:
> @bachradsusi is this PR ok to merge?

yes
